### PR TITLE
Bogus null contract violation diagnostic at yield statements

### PR DESCRIPTION
**Pull request by srikanth-sankaran**
## What it does
* Fixes https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2522

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by srikanth-sankaran** [comment]
@stephan-herrmann my first contribution to null analysis after 10 years hiatus. Be gentle and kind in review comments 😆

**Comment by stephan-herrmann** [comment]
> `but the provided value is null` reads better as `but the provided value may be null` but that is nitpicking

Is that a linguistic hint? AFAICS there is no doubt about null at the two error locations in the new test, so 'is' is more precise than 'may be'. Do you disagree about this analysis?

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java [hunk 1]
```diff
@@ -55,8 +55,6 @@ public FlowInfo analyseCode(BlockScope currentScope, FlowContext flowContext, Fl
 
 	flowInfo = this.expression.analyseCode(currentScope, flowContext, flowInfo);
 	this.expression.checkNPEbyUnboxing(currentScope, flowContext, flowInfo);
-	if (flowInfo.reachMode() == FlowInfo.REACHABLE && currentScope.compilerOptions().isAnnotationBasedNullAnalysisEnabled)
-		checkAgainstNullAnnotation(currentScope, flowContext, flowInfo, this.expression);
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java:59]: So you successfully removed a bogus analysis for the price that NPE caused by `yield` will not be detected at all, right?

I admit that the current code is only useful in the narrow situation of `return switch(v) { ... yield x; }`

I believe the better course of action would be to identify the expected annotated type to which the yield value will be assigned, without taking the shortcut to the method return type.

**Review by stephan-herrmann — CHANGES_REQUESTED** [review]
Welcome back to null analysis, @srikanth-sankaran :)

You found the location that triggered incorrect behaviour. I hope you find my detailed comment helpful to further improve the change.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java (hunk 1 above)
Comment (srikanth-sankaran) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java:59]: > So you successfully removed a bogus analysis for the price that NPE caused by `yield` will not be detected at all, right?
> 
> I admit that the current code is only useful in the narrow situation of `return switch(v) { ... yield x; }`
>

Hold on, please take a look again! 

I added a new test via a follow up commit that proves that a null injected by `yield` of a `switch` expression that is the directly `return`ed *will be* detected: See `org.eclipse.jdt.core.tests.compiler.regression.NullAnnotationTests21.testIssue2522_2()`

Why would yield require special treatment under return anymore than ternary expression would ? 
 
> I believe the better course of action would be to identify the expected annotated type to which the yield value will be assigned, without taking the shortcut to the method return type.

This happens automatically! Just the same way this is complained against:

```
@NonNull Stuff l = switch (pm) {
					case Stuff s -> s;
					case null -> null; // error: Null type mismatch: required 'PatternMatching.@NonNull Stuff' but the provided value is null
					};
```

So do you have a snippet that shows a case where we would not detect a null injection due to the change here ?? I

**Comment by srikanth-sankaran** [comment]
> > `but the provided value is null` reads better as `but the provided value may be null` but that is nitpicking
> 
> Is that a linguistic hint? AFAICS there is no doubt about null at the two error locations in the new test, so 'is' is more precise than 'may be'. Do you disagree about this analysis?

No, please ignore that comment - I didn't realize we *also* have a variant `but the provided value is inferred as @Nullable` where the expression may be but also may not be null. So we are good.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java (hunk 1 above)
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/YieldStatement.java:59]: Apologies, I forgot about how a switch expression collects result expressions and how various null-checking methods will drill into those result expressions, without the yield expression needing to do anything. The fact that I wrote my previous review comment on a train is only a weak excuse.

**Review by stephan-herrmann — APPROVED** [review]
Thanks, @srikanth-sankaran,  for insisting. After more scrutiny I see no problem with this PR. Please go ahead.

**Pull request merged**